File: code/python/687.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def longestUnivaluePath(self, root: TreeNode) -> int:
        res = 0
        def dfs(node):
            nonlocal res
            if node is None:
                return 0
            length = 0
            # 取左边最长的
            leftL = dfs(node.left)
            # 取右边最长的
            rightL = dfs(node.right)
            # 以当前节点为根的最长
            if node.left is not None and node.right is not None and node.left.val==node.val and node.right.val==node.val:
                res = max(res,leftL+rightL+2)
            # 不以该节点为根，将左右两条链中长的拿去累加
            if node.left is not None and node.val==node.left.val:
                length = leftL+1
            if node.right is not None and node.val==node.right.val:
                length = max(length,rightL+1)
            res = max(res,length)
            return length
        dfs(root)
        if res==0:return 0
        return res

File: code/java/687.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    int ans = 0;    // 结果
    public int longestUnivaluePath(TreeNode root) {
        helper(root);
        return ans;
    }

    // 搜索以root为起点的最长同值路径
    // 要么经过左子树，要么经过右子树
    public int helper(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int maxLength = 0;  // 以root为起点的最长同值路径
        int leftLength = helper(root.left);     // 以root.left为起点的最长同值路径
        int rightLength = helper(root.right);   // 以root.right为起点的最长同值路径
        // 情况2，不需要更新maxLength，但要更新结果
        if (root.left != null && root.right != null &&
            root.val == root.left.val && root.val == root.right.val) {
                ans = Math.max(ans, leftLength + rightLength + 2);
        }
        // 从左右子树中选取最长同值路径
        if (root.left != null && root.val == root.left.val) {
            maxLength = leftLength + 1;
        }
        if (root.right != null && root.val == root.right.val) {
            maxLength = Math.max(maxLength, rightLength + 1);
        }
        // 更新结果
        ans = Math.max(ans, maxLength);
        return maxLength;
    }
}

作者：zui-weng-jiu-xian
链接：https://leetcode-cn.com/problems/longest-univalue-path/solution/zui-chang-tong-zhi-lu-jing-di-gui-bian-li-by-zui-w/
来源：力扣（LeetCode）
著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。